Add a Neo4j extension guide (#3171)
quarkusio/quarkus: docs/src/main/asciidoc/neo4j-guide.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
:neo4j_version: 3.5.6

> = Quarkus - Neo4j

Reviewer: The title has to be the first element after the comment.

Reviewer: You marked it as resolved but the title is still after the attribute. It needs to be before the attribute (otherwise it's not properly considered as a title on the website).

include::./attributes.adoc[]

https://neo4j.com[Neo4j] is a graph database management system developed by Neo4j, Inc.
Neo4j is a native graph database focused not only on the data itself, but especially on the relations between data.

Neo4j offers Cypher, a declarative query language much like SQL.

> Cypher is used to query the Graph, telling the database engine what patterns to match and not how.

Reviewer: Not sure I understand the "and not how"?

Reviewer: I still don't understand what it's supposed to say.

Clients communicate over the so called Bolt protocol with the database.

Neo4j - as the most popular graph database according to DB-Engines ranking - provides a variety of drivers for various languages.

The Quarkus Neo4j extension is based on the official https://github.com/neo4j/neo4j-java-driver[Neo4j Java Driver].

> The extensions provides an instance of the driver configured ready for usage in any Quarkus application.

Reviewer: ```suggestion
The extension provides an instance of the driver configured ready for usage in any Quarkus application.
```

You will be able to issue arbitrary Cypher statements over Bolt with this extension.
Those statements can be simple CRUD statements as well as complex queries, calling graph algorithms and more.

The driver itself is released under the Apache 2.0 license,
while Neo4j itself is available in a GPL3-licensed open-source "community edition",
with online backup and high availability extensions licensed under a closed-source commercial license.

== Programming model

NOTE: The Neo4j extension is based on an alpha version of the Neo4j driver. Some interactions with the driver might change in the future.

> The driver and thus the Quarkus extension supports three different programming models:

Reviewer: ```suggestion
The driver and thus the Quarkus extension support three different programming models:
```
?

* Blocking database access (much like standard JDBC)

> * Asynchronous programming based on JDKs completable futures and related infrastructure

Reviewer: ```suggestion
* Asynchronous programming based on JDK's completable futures and related infrastructure
```
or JDK?

* Reactive programming based on http://www.reactive-streams.org[Reactive Streams]

The reactive programming model is only available when connected against a 4.0+ version of Neo4j.
Reactive programming in Neo4j is fully end-to-end reactive and therefore requires a server that supports backpressure.

In this guide you will learn how to

* Add the Neo4j extension to your project
* Configure the driver
* And how to use the driver to access a Neo4j database

This guide will focus on asynchronous access to Neo4j, as this is ready to use for everyone.
At the end of this guide, there will be a reactive version, which needs however a 4.0 database version.

== The domain

> As with some of the other guise, the application shall manage fruit entities.

Reviewer: ```suggestion
As with some of the other guides, the application shall manage fruit entities.
```

[source,java]
----
package org.acme.neo4j;

public class Fruit {

    public Long id;

    public String name;

    public Fruit() {
        // This is needed for the REST-Easy JSON Binding
    }

    public Fruit(String name) {
        this.name = name;
    }

    public Fruit(Long id, String name) {
        this.id = id;
        this.name = name;
    }
}
----

== Prerequisites

To complete this guide, you need:

* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* an IDE
* Apache Maven 3.5.3+
* Access to a Neo4j Database
* Optional Docker for your system

=== Setup Neo4j

The easiest way to start a Neo4j instance is a locally installed Docker environment.

[source,shell,subs="verbatim,attributes"]
----
docker run --publish=7474:7474 --publish=7687:7687 -e 'NEO4J_AUTH=neo4j/secret' neo4j:{neo4j_version}
----

> This starts a Neo4j instance, that publishes it's Bolt port on `7687` and a webinterface on http://localhost:7474.

Reviewer: ```suggestion
This starts a Neo4j instance, that publishes its Bolt port on `7687` and a web interface on http://localhost:7474.
```

Have a look at the https://neo4j.com/download/?ref=product[download page] for other options to get started with the product itself.

== Creating the Maven project

First, we need a new project. Create a new project with the following command:

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \

>     -DprojectArtifactId=neo4j-quickstart\

Reviewer: ```suggestion
    -DprojectArtifactId=neo4j-quickstart \
```
(just for cuteness)

    -Dextensions="neo4j"
----

It generates:

* the Maven structure
* a landing page accessible on `http://localhost:8080`
* example `Dockerfile` files for both `native` and `jvm` modes
* the application configuration file
* an `org.acme.datasource.GreetingResource` resource
* an associated test

The Neo4j extension has been added already to your pom.
If you have an already created project, the `neo4j` extension can be added to an existing Quarkus project with the `add-extension` command:

[source,shell]
----
mvn quarkus:add-extension -Dextensions="neo4j"
----

Otherwise, you can manually add this to the dependencies section of your `pom.xml` file:

[source,xml]
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-neo4j</artifactId>
</dependency>
----

=== JSON Binding

We will expose `Fruit` instances over HTTP in the JSON format.
Consequently, you also need to add the `quarkus-resteasy-jsonb` extension:

[source,shell]
----

> mvn quarkus:add-extension -Dextensions="resteasy-jsonb"

Reviewer: Let's add it from the start when initializing the project.

It's a Quarkus extension used a lot in the guides so people should be familiar with it.

Reviewer: You marked the conversation resolved but it's still exactly the same as before?

----

== Configuring

The Neo4j driver can be configured with standard Quarkus properties:

[source]
.src/main/resources/application.properties
----
quarkus.neo4j.uri = bolt://localhost:7687
quarkus.neo4j.authentication.username = neo4j

> quarkus.neo4j.authentication.password = secret

Reviewer: Now looks like the good opportunity to ask a question I have: do you plan to add more properties? I suppose you have a connection pool for instance, that would probably need some configuration knobs?

Author: Yes, indeed. 

This is what we have in our Spring Boot Starter
https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/blob/master/neo4j-java-driver-spring-boot-autoconfigure/src/main/java/org/neo4j/driver/springframework/boot/autoconfigure/Neo4jDriverProperties.java

I would not mind porting it in time.

Reviewer: Yes, that would be great. We at least need the pool configuration.

----

You'll recognize the authentication here that you passed on to the docker command above.

Having done that, the driver is ready to use.

== Using the driver

> === General remarks.

Reviewer: ```suggestion
=== General remarks
```

The result of a statement consists usually of one or more `org.neo4j.driver.Record`.
Those records contain arbitrary values, supported by the driver.

> If you return a node in the graph, it will be a `org.neo4j.driver.types.Node`.

Reviewer: ```suggestion
If you return a node of the graph, it will be a `org.neo4j.driver.types.Node`.
```
?

[source,java]
.src/main/java/org/acme/neo4j/FruitResource.java
----
package org.acme.neo4j;

import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.neo4j.driver.Driver;

@Path("fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class FruitResource {

    @Inject
    Driver driver;
}
----

=== Reading nodes

Add the following method to the fruit resource:

[source,java]
.src/main/java/org/acme/neo4j/FruitResource.java
----
@GET
public CompletionStage<Response> get() {
    AsyncSession session = driver.asyncSession(); // <1>
    return session
        .runAsync("MATCH (f:Fruit) RETURN f ORDER BY f.name")  // <2>
        .thenCompose(cursor ->  // <3>
            cursor.listAsync(record -> Fruit.from(record.get("f").asNode()))
        )
        .thenCompose(fruits ->  // <4>
            session.closeAsync().thenApply(signal -> fruits)
        )
        .thenApply(Response::ok) // <5>
        .thenApply(ResponseBuilder::build);
}
----
<1> Open a new, asynchronous session with Neo4j

> <2> Execute a query. This is a Cypher statement.

Reviewer: Maybe some link to the documentation on `Cypher`?

Author: Added to the top. Very good point, because I think Cypher is a great language, not only for our product. It will be available in Apache Spark and is an inspiration for the https://www.gqlstandards.org GQL effort

> <3> Retrieve a cursor, list the results and create `Fruits`.

Reviewer: ```suggestion
<3> Retrieve a cursor, list the results and create ``Fruit``s.
```

<4> Close the session after processing
<5> Create a JAX-RS response

Now start Quarkus in `dev` mode with:

[source, shell]
----
mvn compile quarkus:dev
----

> and retrieve the endpoints like this

Reviewer: ```suggestion
and retrieve the endpoint like this
```

[source, shell]
----
curl localhost:8080/fruits
----

> There are not any fruits, so create some:

Reviewer: ```suggestion
There are not any fruits, so let's create some.
```

=== Creating nodes

The `POST` method looks similar.
It uses transaction functions of the driver:

[source,java]
.src/main/java/org/acme/neo4j/FruitResource.java
----
@POST
public CompletionStage<Response> create(Fruit fruit) {
    AsyncSession session = driver.asyncSession();
    return session
        .writeTransactionAsync(tx ->
            tx.runAsync("CREATE (f:Fruit {name: $name}) RETURN f", Values.parameters("name", fruit.name))
        )
        .thenCompose(fn -> fn.singleAsync())
        .thenApply(record -> Fruit.from(record.get("f").asNode()))
        .thenCompose(persistedFruid -> session.closeAsync().thenApply(signal -> persistedFruid))
        .thenApply(persistedFruid -> Response
            .created(URI.create("/fruits/" + persistedFruid.id))
            .build()
        );
}
----

As you can see, we are now using a Cypher statement with named parameters (The `$name` of the fruit).

> The node is returned, a fruit entity created and than mapped to a `201` created response.

Reviewer: ```suggestion
The node is returned, a `Fruit` entity created and then mapped to a `201` created response.
```

A curl request against this path may look like this:

[source, shell]
----
curl -v -X "POST" "http://localhost:8080/fruits" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "name": "Banana"
}'
----

The response contains an URI that shall return single nodes.

=== Read single nodes

This time, we signal a read only transaction.

> This way, a Neo4j cluster could select a follower.

Reviewer: `follower`?

Author: Good point. I would have to go into more detail here as it would fit into the guide.

We also add some exception handling, in case the resource is called with an invalid id:

[source,java]
.src/main/java/org/acme/neo4j/FruitResource.java
----
@GET
@Path("{id}")
public CompletionStage<Response> getSingle(@PathParam("id") Long id) {
    AsyncSession session = driver.asyncSession();
    return session
        .readTransactionAsync(tx ->
            tx.runAsync("MATCH (f:Fruit) WHERE id(f) = $id RETURN f", Values.parameters("id", id))
    )
    .thenCompose(fn -> fn.singleAsync())
    .handle((record, exception) -> {
        if(exception != null) {
            Throwable source = exception;
            if(exception instanceof CompletionException) {
                source = ((CompletionException)exception).getCause();
            }
            Status status = Status.INTERNAL_SERVER_ERROR;
            if(source instanceof NoSuchRecordException) {
                status = Status.NOT_FOUND;
            }
            return Response.status(status).build();
        } else  {
            return Response.ok(Fruit.from(record.get("f").asNode())).build();
        }
    })
    .thenCompose(response -> session.closeAsync().thenApply(signal -> response));
}
----

> A request may look like this

Reviewer: ```suggestion
A request may look like this:
```

[source, shell]
----
curl localhost:8080/fruits/42
----

=== Deleting nodes

Finally, we want to get rid of fruits again and we add the `DELETE` method:

[source,java]
.src/main/java/org/acme/neo4j/FruitResource.java
----
@DELETE
@Path("{id}")
public CompletionStage<Response> delete(@PathParam("id") Long id) {

    AsyncSession session = driver.asyncSession();
    return session
        .writeTransactionAsync(tx ->
            tx.runAsync("DELETE (f:Fruit) WHERE id(f) = $id", Values.parameters("id", id))
        )
        .thenCompose(fn -> fn.consumeAsync()) // <1>
        .thenCompose(response -> session.closeAsync())
        .thenApply(signal -> Response.noContent().build());
}
----
<1> There is no result for us, only a summary of the query executed.

A request may look like this

[source, shell]
----
curl -X DELETE localhost:8080/fruits/42
----

And that's already the most simple CRUD application with one type of nodes.
Feel free to add relationships to the model.
One idea would be to model recipes that contain fruits.

> We add the super simple UI of another quick start guide.

Reviewer: In the other guides, we usually skip the UI part as it's not very interesting in the Quarkus context.

See https://quarkus.io/guides/hibernate-search-guide#creating-a-frontend for instance.

I won't review this part for now. If you disagree, I'll review it again.

Author: It's basically a copy of the UI of the reactive Postgres guide. I think it's ok to leave it in the the actual getting started project, but I'll remove it from here.

Reviewer: The problem is not with having an UI. The problem is that we really don't need to describe it and enter into the details.

See the example I pointed out: we just say "go grab the UI from GitHub" and that's OK for this type of guide where we don't really care about frontend development.

But sure, having a UI is nice.

            list = list
                + '<tr>'
                + '<td>' + fruit.id + '</td>'
                + '<td>' + fruit.name + '</td>'
                + '<td><a href="#" onclick="deleteFruit(' + fruit.id + ')">Delete</a></td>'
                + '</tr>'
        });
        if (list.length > 0) {
            list = ''
                + '<table><thead><th>Id</th><th>Name</th><th></th></thead>'
                + list
                + '</table>';
        } else {
            list = "No fruits in database"
        }
        $('#all-fruits').html(list);
    });
}

function deleteFruit(id) {
    $.ajax('/fruits/' + id, {method: 'DELETE'}).then(refresh);
}

$(document).ready(function () {

    $('#create-fruit-button').click(function () {
        var fruitName = $('#fruit-name').val();
        $.post({
            url: '/fruits',
            contentType: 'application/json',
            data: JSON.stringify({name: fruitName})
        }).then(refresh);
    });

    refresh();
});
----
<1> The `fruits` parameter is not defined when the database is empty.

All done!
Navigate to http://localhost:8080/fruits.html and read/create/delete some fruits.

== Next steps

=== Packaging

> Package your application as simple as `./mvnw clean package`.

Reviewer: ```suggestion
Packaging your application is as simple as `./mvnw clean package`.
```

It can be run with `java -jar target/neo4j-quickstart-1.0-SNAPSHOT-runner.jar`.

> With the GraalVM installed, you can also create a native executable binary: `./mvnw clean package -Dnative`.

Reviewer: ```suggestion
With GraalVM installed, you can also create a native executable binary: `./mvnw clean package -Dnative`.
```
